File: src/main/java/math/cat/N.java
package math.cat;

import java.util.AbstractSet;
import java.util.Iterator;

/**
 * Natural numbers class.
 *
 * @author Vlad Patryshev
 * All source code is stored on <a href="http://code.google.com/p/categories/">http://code.google.com/p/categories/</a>
 * 
 */
public class N extends AbstractSet<Integer> {
  @Override
  public Iterator<Integer> iterator() {
    return new Iterator<Integer>() {
      int n = 0;

      public boolean hasNext() {
        return true;
      }

      public Integer next() {
        return n++;
      }

      public void remove() {
        throw new UnsupportedOperationException();
      }
    };
  }

  @Override
  public int size() {
    return Integer.MAX_VALUE;
  }
}

File: src/main/scala/math/cat/N.scala
package math.cat

/**
 * Set of natural numbers.
 */
object N extends BigSet[BigInt] with EnumerableSet[BigInt] {
  override def toString: String = "ℕ"
  
  override def iterator: Iterator[BigInt] = new Iterator[BigInt] {
    private var n: BigInt = -1
  
    override def hasNext: Boolean = true

    private val tooMany: Int = 1000
    
    override def next(): BigInt = {
      n += 1
//      if (n > tooMany) { // this is good for testing
//        throw new IllegalStateException("fuck...")
//      }
      n
    }
  }
  override def contains(n: BigInt): Boolean = n >= 0
}
